PROTOCOL.md

```diff
@@ -1,183 +1,374 @@
+- **Message Layer:** Some messages in the Control Layer carry messages published in streams. The Message Layer defines the format of these message payloads, consisting of data and metadata of the messages.
+
+This documentation describes the Control Layer and Message Layer since they are common to any network configuration.
+
+## Table of contents
+
+- [Control Layer](#control-layer)
+    - [PublishRequest](#publishrequest)
+    - [SubscribeRequest](#subscriberequest)
+    - [UnsubscribeRequest](#unsubscriberequest)
+    - [ResendLastRequest](#resendlastrequest)
+    - [ResendFromRequest](#resendfromrequest)
+    - [ResendRangeRequest](#resendrangerequest)
+    - [BroadcastMessage](#broadcastmessage)
+    - [UnicastMessage](#unicastmessage)
+    - [SubscribeResponse](#subscriberesponse)
+    - [UnsubscribeResponse](#unsubscriberesponse)
+    - [ResendResponseResending](#resendresponseresending)
+    - [ResendResponseResent](#resendresponseresent)
+    - [ResendResponseNoResend](#resendresponsenoresend)
+    - [ErrorResponse](#errorresponse)
+- [Message Layer](#message-layer)
+    - [StreamMessage](#streammessage)
+    - [MessageID](#messageid)
+    - [MessageRef](#messageref)
+
+
+## Control Layer
+
+All messages of the control layer are transmitted as JSON arrays with the following fields : `[version, type, ...typeSpecificFields]`.
+`version` describes the version of the Control Layer. `type` is an integer to identify the message type according to the following table: 
+
+messageType | Description
+----------- | -----------
+0 | BroadcastMessage
+1 | UnicastMessage
+2 | SubscribeResponse
+3 | UnsubscribeResponse
+4 | ResendResponseResending
+5 | ResendResponseResent
+6 | ResendResponseNoResend
+7 | ErrorResponse
+8 | PublishRequest
+9 | SubscribeRequest
+10 | UnsubscribeRequest
+11 | ResendRequest
+
+We start by describing the requests and then the responses.
+
+### Requests sent
 
 Also see the [Javascript client](https://github.com/streamr-dev/streamr-client) documentation.
 
-#### publish
+#### PublishRequest
 
-Publishes a new message to a stream. Requires a write permission to the stream.
+Publishes a new message to a stream. Requires a write permission to the stream. Authentication requires either the API key or the session token to be set. It contains a `StreamMessage` as a payload at the Message Layer level. The `StreamMessage` representation is also an array (nested in the `PublishRequest` array) which is described in the [StreamMessage](#streammessage) section.
 
 ```
-{
-  "type": "publish",
-  "stream": "streamId",
-  "authKey": "authKey",
-  "msg": "{}",                 // the message as stringified json
-  "ts": 1533924184016,         // timestamp (optional), defaults to current time on server
-  "pkey": "deviceId"           // partition key (optional), defaults to none (random partition)
-}
+[version, type, streamMessage, sessionToken, apiKey]
```

Comment:
I think the goal is to get rid of API keys at some point. That's why it's the last parameter. If this is the next change, it will be done in version 2 of the Control Layer.

Reply:
This got me thinking: _What if_ we drop `apiKey` from this new protocol version? We won't be able to completely get rid of API keys yet of course, but is there any reason why the websocket protocol in particular needs to continue to support them? The user could easily use an API key to obtain a `sessionToken`, then use the protocol.